File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/view/ViewPagerPageScrollStateChangedObservable.java
package com.jakewharton.rxbinding2.support.v4.view;

import android.support.v4.view.ViewPager;
import android.support.v4.view.ViewPager.OnPageChangeListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class ViewPagerPageScrollStateChangedObservable extends Observable<Integer> {
  private final ViewPager view;

  ViewPagerPageScrollStateChangedObservable(ViewPager view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Integer> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnPageChangeListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnPageChangeListener {
    private final ViewPager view;
    private final Observer<? super Integer> observer;

    Listener(ViewPager view, Observer<? super Integer> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override public void onPageSelected(int position) {

    }

    @Override public void onPageScrollStateChanged(int state) {
      if (!isDisposed()) {
        observer.onNext(state);
      }
    }

    @Override protected void onDispose() {
      if (!isDisposed()) {
        view.removeOnPageChangeListener(this);
      }
    }
  }
}

File: rxbinding-viewpager/src/main/java/com/jakewharton/rxbinding4/viewpager/ViewPagerPageScrollStateChangedObservable.kt
@file:JvmName("RxViewPager")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.viewpager

import androidx.annotation.CheckResult
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of scroll state change events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun ViewPager.pageScrollStateChanges(): Observable<Int> {
  return ViewPagerPageScrollStateChangedObservable(this)
}

private class ViewPagerPageScrollStateChangedObservable(
  private val view: ViewPager
) : Observable<Int>() {

  override fun subscribeActual(observer: Observer<in Int>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnPageChangeListener(listener)
  }

  private class Listener(
    private val view: ViewPager,
    private val observer: Observer<in Int>
  ) : MainThreadDisposable(), OnPageChangeListener {

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }

    override fun onPageSelected(position: Int) {
    }

    override fun onPageScrollStateChanged(state: Int) {
      if (!isDisposed) {
        observer.onNext(state)
      }
    }

    override fun onDispose() {
      view.removeOnPageChangeListener(this)
    }
  }
}
